File: mkdocs/overrides/devbook/transactions/account-metadata.mjs
import { PrivateKey } from 'symbol-sdk';
import {
	metadataGenerateKey,
	metadataUpdateValue,
	models,
	NetworkTimestamp,
	SymbolFacade
} from 'symbol-sdk/symbol';

const NODE_URL = process.env.NODE_URL ||
	'https://reference.symboltest.net:3001';
console.log('Using node', NODE_URL);

// Helper function to announce a transaction
async function announceTransaction(payload, label) {
	console.log(`Announcing ${label} to /transactions`);
	const response = await fetch(`${NODE_URL}/transactions`, {
		method: 'PUT',
		headers: { 'Content-Type': 'application/json' },
		body: payload
	});
	console.log('  Response:', await response.text());
}

// Helper function to wait for transaction confirmation
async function waitForConfirmation(transactionHash, label) {
	console.log(`Waiting for ${label} confirmation...`);
	for (let attempt = 0; attempt < 60; attempt++) {
		await new Promise(resolve => setTimeout(resolve, 1000));
		try {
			const response = await fetch(
				`${NODE_URL}/transactionStatus/${transactionHash}`);
			const status = await response.json();
			console.log('  Transaction status:', status.group);
			if (status.group === 'confirmed') {
				console.log(`${label} confirmed in`, attempt, 'seconds');
				return;
			}
			if (status.group === 'failed') {
				throw new Error(`${label} failed: ${status.code}`);
			}
		} catch (e) {
			if (e.message.includes('failed'))
				throw e;
			console.log('  Transaction status: unknown');
		}
	}
	throw new Error(`${label} not confirmed after 60 seconds`);
}

const SIGNER_PRIVATE_KEY = process.env.SIGNER_PRIVATE_KEY || (
	'0000000000000000000000000000000000000000000000000000000000000000');
const signerKeyPair = new SymbolFacade.KeyPair(
	new PrivateKey(SIGNER_PRIVATE_KEY));

const facade = new SymbolFacade('testnet');
const signerAddress = facade.network.publicKeyToAddress(
	signerKeyPair.publicKey);
console.log('Signer address:', signerAddress.toString());

try {
	// Fetch current network time
	const timePath = '/node/time';
	console.log('Fetching current network time from', timePath);
	const timeResponse = await fetch(`${NODE_URL}${timePath}`);
	const timeJSON = await timeResponse.json();
	const timestamp = new NetworkTimestamp(
		timeJSON.communicationTimestamps.receiveTimestamp);
	console.log('  Network time:', timestamp.timestamp,
		'ms since nemesis');

	// Fetch recommended fees
	const feePath = '/network/fees/transaction';
	console.log('Fetching recommended fees from', feePath);
	const feeResponse = await fetch(`${NODE_URL}${feePath}`);
	const feeJSON = await feeResponse.json();
	const medianMult = feeJSON.medianFeeMultiplier;
	const minimumMult = feeJSON.minFeeMultiplier;
	const feeMult = Math.max(medianMult, minimumMult);
	console.log('  Fee multiplier:', feeMult);

	// Define metadata key and value
	const keyString = `username_${Date.now()}`;
	const scopedMetadataKey = metadataGenerateKey(keyString);
	const metadataValue = new TextEncoder().encode('alice');

	// Create the embedded metadata transaction
	const embeddedTransaction = facade.transactionFactory
		.createEmbedded({
			type: 'account_metadata_transaction_v1',
			signerPublicKey: signerKeyPair.publicKey.toString(),
			targetAddress: signerAddress.toString(),
			scopedMetadataKey,
			// When creating new metadata, valueSizeDelta
			// equals value length
			valueSizeDelta: metadataValue.length,
			value: metadataValue
		});
	console.log('Created embedded metadata transaction:');
	console.log(JSON.stringify(embeddedTransaction.toJson(), null, 2));

	// Build the aggregate transaction
	const embeddedTransactions = [embeddedTransaction];
	const transaction = facade.transactionFactory.create({
		type: 'aggregate_complete_transaction_v3',
		signerPublicKey: signerKeyPair.publicKey.toString(),
		deadline: timestamp.addHours(2).timestamp,
		transactionsHash: facade.static.hashEmbeddedTransactions(
			embeddedTransactions),
		transactions: embeddedTransactions
	});
	transaction.fee = new models.Amount(feeMult * transaction.size);
	console.log('Built aggregate transaction:');
	console.log(JSON.stringify(transaction.toJson(), null, 2));

	// Sign and generate final payload
	const signature = facade.signTransaction(signerKeyPair, transaction);
	const jsonPayload = facade.transactionFactory.static.attachSignature(
		transaction, signature);

	// Announce and wait for confirmation
	const transactionHash =
		facade.hashTransaction(transaction).toString();
	await announceTransaction(jsonPayload, 'aggregate transaction');
	await waitForConfirmation(transactionHash, 'Aggregate transaction');

	// --- MODIFYING EXISTING METADATA ---
	console.log('\n--- Modifying existing metadata ---');

	// Fetch current metadata value from network
	const metadataPath = `/metadata?sourceAddress=${signerAddress}`
		+ `&targetAddress=${signerAddress}`
		+ '&metadataType=0';
	console.log('Fetching current metadata from', metadataPath);
	const metadataResponse = await fetch(`${NODE_URL}${metadataPath}`);
	const metadataJSON = await metadataResponse.json();

	// Find the entry matching our scoped metadata key
	const foundEntry = metadataJSON.data.find(
		entry => BigInt(`0x${entry.metadataEntry.scopedMetadataKey}`)
			=== scopedMetadataKey
	);
	if (!foundEntry) {
		throw new Error('Metadata entry not found');
	}
	const metadataEntry = foundEntry.metadataEntry;
	const currentValue = Buffer.from(metadataEntry.value, 'hex');
	console.log('  Current value:', currentValue.toString('utf8'));

	// XOR the current and new values
	const newValue = new TextEncoder().encode('bob');
	const updateValue = metadataUpdateValue(currentValue, newValue);

	// Create the update transaction with XOR'd value
	const embeddedUpdate = facade.transactionFactory
		.createEmbedded({
			type: 'account_metadata_transaction_v1',
			signerPublicKey: signerKeyPair.publicKey.toString(),
			targetAddress: signerAddress.toString(),
			scopedMetadataKey,
			// valueSizeDelta is the difference in length
			// (can be negative)
			valueSizeDelta: newValue.length - currentValue.length,
			value: updateValue
		});

	// Build the aggregate for the update
	const updateEmbedded = [embeddedUpdate];
	const updateTransaction = facade.transactionFactory.create({
		type: 'aggregate_complete_transaction_v3',
		signerPublicKey: signerKeyPair.publicKey.toString(),
		deadline: timestamp.addHours(2).timestamp,
		transactionsHash: facade.static.hashEmbeddedTransactions(
			updateEmbedded),
		transactions: updateEmbedded
	});
	updateTransaction.fee = new models.Amount(
		feeMult * updateTransaction.size);
	console.log('Built aggregate transaction:');
	console.log(JSON.stringify(updateTransaction.toJson(), null, 2));

	// Sign and announce the update
	const updateSignature = facade.signTransaction(
		signerKeyPair, updateTransaction);
	const updatePayload = facade.transactionFactory.static
		.attachSignature(updateTransaction, updateSignature);

	// Announce and wait for confirmation
	const updateHash =
		facade.hashTransaction(updateTransaction).toString();
	await announceTransaction(updatePayload, 'aggregate transaction');
	await waitForConfirmation(updateHash, 'Aggregate transaction');

} catch (e) {
	console.error(e.message, '| Cause:', e.cause?.code ?? 'unknown');
}

File: mkdocs/snippets/devbook/accounts/AccountMetadata.java
//JAVA 21+
//DEPS org.symbol:symbol-sdk:3.3.1

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.http.HttpResponse.BodyHandlers;
import java.nio.charset.StandardCharsets;
import java.util.List;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import org.symbol.sdk.CryptoTypes;
import org.symbol.sdk.facade.SymbolFacade;
import org.symbol.sdk.symbol.Address;
import org.symbol.sdk.symbol.KeyPair;
import org.symbol.sdk.symbol.Metadata;
import org.symbol.sdk.symbol.SymbolTransactionFactory;
import org.symbol.sdk.symbol.descriptors.*;
import org.symbol.sdk.symbol.models.*;

public final class AccountMetadata {
	private static final ObjectMapper JSON_MAPPER = new ObjectMapper();

	private static final HttpClient HTTP_CLIENT =
		HttpClient.newHttpClient();

	private final String nodeUrl = System.getenv().getOrDefault(
		"NODE_URL", "https://reference.symboltest.net:3001");

	private final SymbolFacade facade = new SymbolFacade("testnet");

	private void announceTransaction(
		final String payload,
		final String label
	) throws IOException, InterruptedException {
		System.out.printf("Announcing %s to /transactions%n", label);
		final HttpRequest request = HttpRequest.newBuilder(
			URI.create(nodeUrl + "/transactions"))
			.header("Content-Type", "application/json")
			.PUT(HttpRequest.BodyPublishers.ofString(payload))
			.build();
		final HttpResponse<String> response = HTTP_CLIENT.send(
			request, BodyHandlers.ofString());
		System.out.printf("  Response: %s%n", response.body());
	}

	private void waitForConfirmation(
		final String transactionHash,
		final String label
	) throws IOException, InterruptedException {
		System.out.printf("Waiting for %s confirmation...%n", label);
		for (int attempt = 0; 60 > attempt; ++attempt) {
			Thread.sleep(1000);
			final String statusPath =
				"/transactionStatus/" + transactionHash;
			final HttpRequest statusRequest = HttpRequest.newBuilder(
				URI.create(nodeUrl + statusPath)).GET().build();
			final HttpResponse<String> statusResponse = HTTP_CLIENT
				.send(statusRequest, BodyHandlers.ofString());
			if (404 == statusResponse.statusCode()) {
				System.out.println("  Transaction status: unknown");
				continue;
			}
			if (2 != statusResponse.statusCode() / 100)
				throw new IOException(
					"HTTP " + statusResponse.statusCode());

			final JsonNode status =
				JSON_MAPPER.readTree(statusResponse.body());
			final String group = status.get("group").asText();
			System.out.printf("  Transaction status: %s%n", group);
			if ("confirmed".equals(group)) {
				System.out.printf("%s confirmed in %d seconds%n",
					label, attempt);
				return;
			}
			if ("failed".equals(group))
				throw new IOException(String.format("%s failed: %s",
					label, status.get("code").asText()));
		}
		throw new IOException(String.format(
			"%s not confirmed after 60 seconds", label));
	}

	public static void main(final String[] args) {
		try {
			new AccountMetadata().run();
		} catch (final Exception ex) {
			System.out.println(null == ex.getMessage()
				? ex.toString()
				: ex.getMessage());
		}
	}

	private void run() throws IOException, InterruptedException {
		System.out.printf("Using node %s%n", nodeUrl);

		// [>step-1]
		final String privateKeyString = System.getenv().getOrDefault(
			"SIGNER_PRIVATE_KEY", "0".repeat(64));
		final KeyPair signerKeyPair = new KeyPair(
			new CryptoTypes.PrivateKey(privateKeyString));

		final Address signerAddress = facade.network.publicKeyToAddress(
			signerKeyPair.getPublicKey());
		System.out.printf("Signer address: %s%n", signerAddress);
		// [<step-1]
		// Fetch recommended fees [>step-2]
		final String feePath = "/network/fees/transaction";
		System.out.printf("Fetching recommended fees from %s%n", feePath);
		final HttpRequest feeRequest = HttpRequest.newBuilder(
			URI.create(nodeUrl + feePath)).GET().build();
		final HttpResponse<String> feeResponse = HTTP_CLIENT.send(
			feeRequest, BodyHandlers.ofString());
		final JsonNode feeJson = JSON_MAPPER.readTree(feeResponse.body());
		final long feeMultiplier = Math.max(
			feeJson.get("medianFeeMultiplier").asLong(),
			feeJson.get("minFeeMultiplier").asLong());
		System.out.printf("  Fee multiplier: %d%n", feeMultiplier);
		// [<step-2]
		System.out.println("\n--- Adding new metadata ---");

		// Define metadata key and value [>step-3]
		final String keyString = String.format(
			"username_%d", System.currentTimeMillis());
		final long scopedMetadataKey = Metadata.generateKey(keyString);
		final byte[] metadataValue =
			"alice".getBytes(StandardCharsets.UTF_8);
		// [<step-3]
		// Create the embedded metadata transaction [>step-4]
		final EmbeddedTransaction creationEmbeddedTx =
			facade.createEmbeddedTransactionFromTypedDescriptor(
				new AccountMetadataTransactionV1Descriptor(
					signerAddress,
					scopedMetadataKey,
					metadataValue.length,
					metadataValue),
				signerKeyPair.getPublicKey());
		System.out.println("Created embedded metadata transaction:");
		System.out.println(JSON_MAPPER
			.writerWithDefaultPrettyPrinter()
			.writeValueAsString(creationEmbeddedTx.toJson()));
		// [<step-4]
		// Build the aggregate transaction [>step-5]
		final List<EmbeddedTransaction> creationEmbeddedTxs =
			List.of(creationEmbeddedTx);
		final Transaction creationTx =
			facade.createTransactionFromTypedDescriptor(
				new AggregateCompleteTransactionV3Descriptor(
					SymbolFacade.hashEmbeddedTransactions(
						creationEmbeddedTxs),
					creationEmbeddedTxs,
					null),
				signerKeyPair.getPublicKey(),
				feeMultiplier,
				2 * 60 * 60);
		// [<step-5]
		// Sign and generate final payload [>step-6]
		final String creationPayload = SymbolTransactionFactory
			.attachSignature(creationTx,
				facade.signTransaction(signerKeyPair, creationTx));

		// Announce and wait for confirmation
		final String creationTxHash =
			facade.hashTransaction(creationTx).toString();
		System.out.printf(
			"Built aggregate transaction with hash: %s%n",
			creationTxHash);
		announceTransaction(creationPayload, "creation transaction");
		waitForConfirmation(creationTxHash, "creation transaction");
		// [<step-6]
		System.out.println("\n--- Modifying existing metadata ---");

		// Fetch current metadata value from network [>step-7]
		final String scopedKeyHex = "%016X".formatted(
			scopedMetadataKey);
		final String metadataPath = String.format(
			"/metadata?sourceAddress=%s&targetAddress=%s"
				+ "&scopedMetadataKey=%s&metadataType=0",
			signerAddress, signerAddress, scopedKeyHex);
		System.out.printf("Fetching current metadata from %s%n",
			metadataPath);
		final HttpRequest metadataRequest = HttpRequest.newBuilder(
			URI.create(nodeUrl + metadataPath)).GET().build();
		final HttpResponse<String> metadataResponse = HTTP_CLIENT.send(
			metadataRequest, BodyHandlers.ofString());
		final JsonNode metadataJson = JSON_MAPPER.readTree(
			metadataResponse.body());

		// Get the metadata entry
		final JsonNode metadataData = metadataJson.get("data");
		if (null == metadataData)
			throw new IOException(
				"Unexpected metadata response: " + metadataJson);

		if (metadataData.isEmpty())
			throw new IOException("Metadata entry not found");

		final JsonNode metadataEntry = metadataData.get(0)
			.get("metadataEntry");
		final byte[] currentValue = java.util.HexFormat.of()
			.parseHex(metadataEntry.get("value").asText());
		System.out.printf("  Current value: %s%n",
			new String(currentValue, StandardCharsets.UTF_8));
		// [<step-7]
		// XOR the current and new values [>step-8]
		final byte[] newValue = "bob".getBytes(StandardCharsets.UTF_8);
		final byte[] updateValue = Metadata.updateValue(
			currentValue, newValue);

		// Create the update transaction with XOR'd value
		final EmbeddedTransaction updateEmbeddedTx =
			facade.createEmbeddedTransactionFromTypedDescriptor(
				new AccountMetadataTransactionV1Descriptor(
					signerAddress,
					scopedMetadataKey,
					newValue.length - currentValue.length,
					updateValue),
				signerKeyPair.getPublicKey());
		// [<step-8]
		// Build the aggregate for the update [>step-9]
		final List<EmbeddedTransaction> updateEmbeddedTxs =
			List.of(updateEmbeddedTx);
		final Transaction updateTx =
			facade.createTransactionFromTypedDescriptor(
				new AggregateCompleteTransactionV3Descriptor(
					SymbolFacade.hashEmbeddedTransactions(
						updateEmbeddedTxs),
					updateEmbeddedTxs,
					null),
				signerKeyPair.getPublicKey(),
				feeMultiplier,
				2 * 60 * 60);

		// Sign and announce the update
		final String updatePayload = SymbolTransactionFactory
			.attachSignature(updateTx,
				facade.signTransaction(signerKeyPair, updateTx));

		// Announce and wait for confirmation
		final String updateTxHash =
			facade.hashTransaction(updateTx).toString();
		System.out.printf(
			"Built aggregate transaction with hash: %s%n",
			updateTxHash);
		announceTransaction(updatePayload, "update transaction");
		waitForConfirmation(updateTxHash, "update transaction");
		// [<step-9]
	}
}
